### PR TITLE
[Spark] Support setting different resources for executor and driver

### PR DESCRIPTION
**Pull request by urihoenig**

### CODE REVIEW

File: mlrun/runtimes/sparkjob.py [hunk 1]
```diff
@@ -161,10 +162,9 @@ def _run(self, runobj: RunObject, execution: MLClientCtx):
         if self.spec.image:
             update_in(job, "spec.image", self.spec.image)
         elif config.igz_version:
-            image_tag = (
-                "" if ":" in config.spark_app_image else ":" + config.igz_version
+            update_in(
+                job, "spec.image", config.spark_app_image + ":" + config.igz_version
```
Comment (Hedingber) [inline mlrun/runtimes/sparkjob.py:166]: Why ?

File: mlrun/runtimes/sparkjob.py [hunk 2]
```diff
@@ -177,19 +177,63 @@ def _run(self, runobj: RunObject, execution: MLClientCtx):
...
                     job, "spec.executor.cores", self.spec.resources["requests"]["cpu"]
                 )
-        if "limits" in self.spec.resources:
-            if "cpu" in self.spec.resources["limits"]:
+            if "memory" in self.spec.resources["requests"]:
                 update_in(
-                    job, "spec.executor.coreLimit", self.spec.resources["limits"]["cpu"]
+                    job,
+                    "spec.executor.memory",
+                    self.spec.resources["requests"]["memory"],
```
Comment (Hedingber) [inline mlrun/runtimes/sparkjob.py:194]: Are you sure the spark format equals the k8s format ?

File: mlrun/runtimes/sparkjob.py [hunk 3]
```diff
@@ -177,19 +177,63 @@ def _run(self, runobj: RunObject, execution: MLClientCtx):
...
+                )
+            gpu_type = [
+                x for x in self.spec.resources["requests"] if x not in ["cpu", "memory"]
```
Comment (Hedingber) [inline mlrun/runtimes/sparkjob.py:197]:
```suggestion
                resource_type for resource_type in self.spec.resources["requests"].keys() if resource_type not in ["cpu", "memory"]
```

File: mlrun/runtimes/sparkjob.py [hunk 4]
```diff
@@ -177,19 +177,63 @@ def _run(self, runobj: RunObject, execution: MLClientCtx):
...
+            ]
+            if gpu_type:
+                update_in(job, "spec.executor.gpu.name", gpu_type[0])
```
Comment (Hedingber) [inline mlrun/runtimes/sparkjob.py:200]: If for some reason more than one type of gpu was added, you're simply ignoring it
please add an `if len(gpu_type) > 1` that will log some warning so at least we will be able to debug this scenario

File: mlrun/runtimes/sparkjob.py [hunk 5]
```diff
@@ -177,19 +177,63 @@ def _run(self, runobj: RunObject, execution: MLClientCtx):
...
+                    "spec.executor.gpu.quantity",
+                    self.spec.resources["requests"][gpu_type[0]],
+                )
+        if "limits" in self.spec.driver_resources:
+            if "cpu" in self.spec.driver_resources["limits"]:
+                update_in(
+                    job,
+                    "spec.driver.coreLimit",
+                    self.spec.driver_resources["limits"]["cpu"],
+                )
```
Comment (Hedingber) [inline mlrun/runtimes/sparkjob.py:212]: No support for memory limit for drivers ?

File: mlrun/runtimes/sparkjob.py [hunk 6]
```diff
@@ -177,19 +177,63 @@ def _run(self, runobj: RunObject, execution: MLClientCtx):
...
+                    "spec.driver.memory",
+                    self.spec.driver_resources["requests"]["memory"],
+                )
+            gpu_type = [
+                x
+                for x in self.spec.driver_resources["requests"]
+                if x not in ["cpu", "memory"]
+            ]
+            if gpu_type:
+                update_in(job, "spec.driver.gpu.name", gpu_type[0])
```
Comment (Hedingber) [inline mlrun/runtimes/sparkjob.py:232]: Same comments regarding the similar code above

File: mlrun/runtimes/sparkjob.py (hunk 5 above)
Comment (Hedingber) [reply mlrun/runtimes/sparkjob.py:212]: also for executor ?

File: mlrun/runtimes/utils.py [hunk 7]
```diff
@@ -341,6 +341,18 @@ def get_resource_labels(function, run=None, scrape_metrics=False):
     return labels
 
 
+def get_resources(mem=None, cpu=None, gpus=None, gpu_type="nvidia.com/gpu"):
```
Comment (Hedingber) [inline mlrun/runtimes/utils.py:344]:
```suggestion
def generate_resources(mem=None, cpu=None, gpus=None, gpu_type="nvidia.com/gpu"):
```

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/runtimes/sparkjob.py (hunk 1 above)
Comment (Hedingber) [reply mlrun/runtimes/sparkjob.py:166]: Now looking on what you did in Provazio I see why you did it,
but please still support the option to give image without tag

File: mlrun/runtimes/sparkjob.py (hunk 4 above)
Comment (urihoenig) [reply mlrun/runtimes/sparkjob.py:200]: I'm ignoring it as SparkApplication support only one (https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/6f35a11526257c3ed6d48b68f48e06d9f1e48999/pkg/apis/sparkoperator.k8s.io/v1beta2/types.go#L407)
+ the with_driver/executor_limits do not allow adding more than one anyway.

The question here is how to fail as this code runs on the server side.
I can throw an exception (not sure if it's the common way to go here).
If I don't want to fail, but only warn, my only way to go here is to use runobj.status.status_text which is already used for the UI URL

File: mlrun/runtimes/sparkjob.py (hunk 5 above)
Comment (urihoenig) [reply mlrun/runtimes/sparkjob.py:212]: No. I think that they duplicate the request to the limit

File: mlrun/runtimes/sparkjob.py (hunk 1 above)
Comment (urihoenig) [reply mlrun/runtimes/sparkjob.py:166]: done

File: mlrun/runtimes/sparkjob.py (hunk 2 above)
Comment (urihoenig) [reply mlrun/runtimes/sparkjob.py:194]: I'm quite sure it's eventually the same (although it goes through spark params), but why should I care?
I'm using a specific interface for sparkjob and there's no input validation any way.

File: mlrun/runtimes/sparkjob.py (hunk 4 above)
Comment (Hedingber) [reply mlrun/runtimes/sparkjob.py:206]: Okay, I see why it basically can't happen, but still please check that for sanity, if length bigger than 1, raise an exception (yes you can do it and it is okay).
Please also use `self.spec.executor_resources` for the executor and don't use `self.spec.resources` for it

Comment (urihoenig) [reply mlrun/runtimes/sparkjob.py:214]: done

**Review by Hedingber — APPROVED** [review]

**Pull request merged**